### PR TITLE
LWDEV-7724 Animated orderbook and tradelog updates

### PR DESCRIPTION
**Pull request by sergeyschegolev**

### CODE REVIEW

File: src/utils/number.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+const START_EXPONENT = 1e-7;
+
+export const getDigits = (x: number) => {
+  if (x <= START_EXPONENT) {
```
Comment (optimusway) [inline src/utils/number.ts]: `Number.EPSILON` fits better here, if i got it right

File: src/utils/__tests__/string.spec.ts [hunk 2]
```diff
@@ -66,10 +68,30 @@ describe('string utils', () => {
...
 
-  it('shoud return string from rest error object', () => {
+  it('should return string from rest error object', () => {
     const restMessage = {id: ['Value to small', 'Please, try again']};
     expect(getRestErrorMessage(restMessage)).toBe(
       'Value to small. Please, try again'
     );
   });
+
+  it('should return true if trailing zero is not present', () => {
```
Comment (optimusway) [inline src/utils/__tests__/string.spec.ts]: plz wrap in `describe yourFunction`

File: src/utils/__tests__/string.spec.ts [hunk 3]
```diff
@@ -66,10 +68,30 @@ describe('string utils', () => {
...
+    const s = '0.12340';
+    expect(hasTrailingZeroes(s)).toBeTruthy();
+  });
+
+  it('should return false if trailing zero is not present', () => {
+    const s = '0.1234';
+    expect(hasTrailingZeroes(s)).toBeFalsy();
+  });
+
+  it('should return start position for trailing zeroes', () => {
```
Comment (optimusway) [inline src/utils/__tests__/string.spec.ts]: is it really the _start_ position? i'd say opposite

File: src/utils/string.ts [hunk 4]
```diff
@@ -47,3 +47,13 @@ export const toLocaleStringWithAccuracy = (
     maximumFractionDigits: accuracy,
     ...options
   });
+
+export const hasTrailingZeroes = (s: string) => s[s.length - 1] === '0';
+
+export const getTrailingZeroStartPosition = (s: string) => {
```
Comment (optimusway) [inline src/utils/string.ts]: could you plz check it works for `getTrailingZeroStartPosition('100.0010')`

File: src/utils/number.ts (hunk 1 above)
Comment (sergeyschegolev) [reply src/utils/number.ts]: Number.EPSILON is too small for this case. The code should catch exactly exponent.

File: src/components/OrderBook/LevelList.tsx [hunk 5]
```diff
@@ -78,17 +108,40 @@ class LevelList extends React.Component<LevelListProps> {
     window.requestAnimationFrame(() => {
       this.renderCanvas(asks, bids, type);
       this.forceUpdate();
+
+      clearInterval(this.cancelColorAnimationIntervalId);
+
+      this.cancelColorAnimationIntervalId = setInterval(() => {
+        if (!!this.animatingLevels.length) {
```
Comment (PaulRumyantsevItr) [inline src/components/OrderBook/LevelList.tsx]: ```!!``` is redundant

File: src/components/OrderBook/LevelList.tsx [hunk 6]
```diff
@@ -136,18 +189,49 @@ class LevelList extends React.Component<LevelListProps> {
       Math.max(...vals)
     );
 
+    const isAnimationPrevented = !this.cachedLevels.length;
+
     levels.forEach((l, i: number) => {
```
Comment (PaulRumyantsevItr) [inline src/components/OrderBook/LevelList.tsx]: Please split code in function by smaller functions

File: src/components/OrderBook/LevelList.tsx [hunk 7]
```diff
@@ -136,18 +189,49 @@ class LevelList extends React.Component<LevelListProps> {
...
+      const existedLevel = this.cachedLevels.find(ml => ml.price === l.price);
```
Comment (PaulRumyantsevItr) [inline src/components/OrderBook/LevelList.tsx]: Why ml? Please use full names, I don't get why cached level has prefix m

File: src/components/OrderBook/LevelList.tsx [hunk 8]
```diff
@@ -199,15 +283,40 @@ class LevelList extends React.Component<LevelListProps> {
...
+        const trailingZeroPosition = getTrailingZeroOppositePosition(volume);
+
+        volume.split('').forEach((symbol: string, index: number) => {
+          const symbolColor =
+            index < trailingZeroPosition ? volumeColor : colors.coolGrey;
+          drawText({
+            ctx: this.canvasCtx,
+            color: symbolColor,
+            text: symbol,
+            x: width / 3 + LEFT_PADDING + drownSymbolsWidth,
```
Comment (PaulRumyantsevItr) [inline src/components/OrderBook/LevelList.tsx]: 3 is magic number here

File: src/components/OrderBook/LevelList.tsx [hunk 9]
```diff
@@ -199,15 +283,40 @@ class LevelList extends React.Component<LevelListProps> {
...
-        color,
-        text: format(l[displayType], instrument.baseAsset.accuracy),
-        x: width / 3 + LEFT_PADDING,
-        y: canvasY - TOP_PADDING,
-        font: LEVEL_FONT,
-        align: 'start'
-      });
+      const volume = format(l[displayType], instrument.baseAsset.accuracy);
+
+      if (hasTrailingZeroes(volume)) {
```
Comment (PaulRumyantsevItr) [inline src/components/OrderBook/LevelList.tsx]: Split by functions

File: src/components/TradeList/PublicTradeList.tsx [hunk 10]
```diff
@@ -1,19 +1,31 @@
...
 }) => (
   <Table>
-    <tbody>
-      {trades.map(trade => <PublicTradeListItem key={trade.id} {...trade} />)}
-    </tbody>
+    <TransitionGroup component={'tbody'}>
+      {trades.map(trade => (
+        <CSSTransition
+          key={trade.id}
+          timeout={300}
```
Comment (PaulRumyantsevItr) [inline src/components/TradeList/PublicTradeList.tsx]: Move to constant

File: src/components/TradeList/styles.css [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+.sell-enter-done {
+  animation: color-sell-out 0.3s alternate;
+}
+
+@keyframes color-sell-in {
+  0% {
+    background-color: transparent;
+  }
+  100% {
+    background-color: #ff6161;
```
Comment (PaulRumyantsevItr) [inline src/components/TradeList/styles.css]: Use colors from ```styled```

File: src/stores/orderBookStore.ts [hunk 12]
```diff
@@ -90,8 +104,14 @@ class OrderBookStore extends BaseStore {
     this.mapToOrderInWorker = this.worker(mapToOrder);
   }
 
-  setAsksUpdatingHandler = (cb: any) => (this.drawAsks = cb);
-  setBidsUpdatingHandler = (cb: any) => (this.drawBids = cb);
+  setAsksUpdatingHandler = (
+    cb: (a: Order[], b: Order[], t: LevelType) => void
```
Comment (PaulRumyantsevItr) [inline src/stores/orderBookStore.ts]: That alphabet in parameters...

File: src/components/OrderBook/helpers/LevelListHelpers.spec.ts [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+
+    orderType = 'depth';
+    expect(getCellType(orderType)).toBe(OrderBookCellType.Depth);
+  });
+
+  it('should return y position for bid', () => {
+    const bidIndex = 1;
+    const side = Side.Buy;
+    const levelHeight = 30;
+    expect(getY(side, bidIndex, levelHeight)).toBe(bidIndex * levelHeight);
```
Comment (PaulRumyantsevItr) [inline src/components/OrderBook/helpers/LevelListHelpers.spec.ts]: Use constant values in toBe parts

File: src/components/OrderBook/styles.ts [hunk 14]
```diff
@@ -89,9 +89,17 @@ export const StyledHeader = Table.extend`
...
     text-align: left;
     padding: 0;
+
+    &:first-child {
+      padding-left: 50px !important;
+    }
   }
 `;
 
+export const FakeTh = styled.th`
```
Comment (PaulRumyantsevItr) [inline src/components/OrderBook/styles.ts]: Not sure why is it fake

File: src/components/styled.ts [hunk 15]
```diff
@@ -89,5 +92,29 @@ export const greyButton = {
 
 export const tableScrollMargin = '1rem';
 
+export let sellRGB = {
+  r: 255,
+  g: 255,
+  b: 255
+};
+
+export let buyRGB = {
```
Comment (PaulRumyantsevItr) [inline src/components/styled.ts]: Is it correct that they are similar?

File: src/utils/string.ts [hunk 16]
```diff
@@ -54,3 +54,17 @@ export const toLocaleStringWithAccuracy = (
...
     ...options
   });
+
+export const hasTrailingZeroes = (s: string) => s[s.length - 1] === '0';
+
+export const getTrailingZeroOppositePosition = (s: string) => {
+  const reversePosition = s
+    .split('')
+    .reverse()
+    .findIndex((smb: string) => smb !== '0');
```
Comment (PaulRumyantsevItr) [inline src/utils/string.ts]: What is s and smb?

**Pull request merged**